// @bnb-chain/greenfield-cosmos-types@0.4.0-alpha.23/greenfield/sp/params.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
//@ts-nocheck
/* eslint-disable */
const helpers_1 = require("../../helpers");
const _m0 = __importStar(require("protobufjs/minimal"));
exports.protobufPackage = "greenfield.sp";
function createBaseParams() {
    return {
        depositDenom: "",
        minDeposit: "",
        secondarySpStorePriceRatio: "",
        numOfHistoricalBlocksForMaintenanceRecords: helpers_1.Long.ZERO,
        maintenanceDurationQuota: helpers_1.Long.ZERO,
        numOfLockupBlocksForMaintenance: helpers_1.Long.ZERO,
        updateGlobalPriceInterval: helpers_1.Long.UZERO,
        updatePriceDisallowedDays: 0
    };
}
exports.Params = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.depositDenom !== "") {
            writer.uint32(10).string(message.depositDenom);
        }
        if (message.minDeposit !== "") {
            writer.uint32(18).string(message.minDeposit);
        }
        if (message.secondarySpStorePriceRatio !== "") {
            writer.uint32(26).string(message.secondarySpStorePriceRatio);
        }
        if (!message.numOfHistoricalBlocksForMaintenanceRecords.isZero()) {
            writer.uint32(32).int64(message.numOfHistoricalBlocksForMaintenanceRecords);
        }
        if (!message.maintenanceDurationQuota.isZero()) {
            writer.uint32(40).int64(message.maintenanceDurationQuota);
        }
        if (!message.numOfLockupBlocksForMaintenance.isZero()) {
            writer.uint32(48).int64(message.numOfLockupBlocksForMaintenance);
        }
        if (!message.updateGlobalPriceInterval.isZero()) {
            writer.uint32(56).uint64(message.updateGlobalPriceInterval);
        }
        if (message.updatePriceDisallowedDays !== 0) {
            writer.uint32(64).uint32(message.updatePriceDisallowedDays);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.depositDenom = reader.string();
                    break;
                case 2:
                    message.minDeposit = reader.string();
                    break;
                case 3:
                    message.secondarySpStorePriceRatio = reader.string();
                    break;
                case 4:
                    message.numOfHistoricalBlocksForMaintenanceRecords = reader.int64();
                    break;
                case 5:
                    message.maintenanceDurationQuota = reader.int64();
                    break;
                case 6:
                    message.numOfLockupBlocksForMaintenance = reader.int64();
                    break;
                case 7:
                    message.updateGlobalPriceInterval = reader.uint64();
                    break;
                case 8:
                    message.updatePriceDisallowedDays = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            depositDenom: (0, helpers_1.isSet)(object.depositDenom) ? String(object.depositDenom) : "",
            minDeposit: (0, helpers_1.isSet)(object.minDeposit) ? String(object.minDeposit) : "",
            secondarySpStorePriceRatio: (0, helpers_1.isSet)(object.secondarySpStorePriceRatio) ? String(object.secondarySpStorePriceRatio) : "",
            numOfHistoricalBlocksForMaintenanceRecords: (0, helpers_1.isSet)(object.numOfHistoricalBlocksForMaintenanceRecords) ? helpers_1.Long.fromValue(object.numOfHistoricalBlocksForMaintenanceRecords) : helpers_1.Long.ZERO,
            maintenanceDurationQuota: (0, helpers_1.isSet)(object.maintenanceDurationQuota) ? helpers_1.Long.fromValue(object.maintenanceDurationQuota) : helpers_1.Long.ZERO,
            numOfLockupBlocksForMaintenance: (0, helpers_1.isSet)(object.numOfLockupBlocksForMaintenance) ? helpers_1.Long.fromValue(object.numOfLockupBlocksForMaintenance) : helpers_1.Long.ZERO,
            updateGlobalPriceInterval: (0, helpers_1.isSet)(object.updateGlobalPriceInterval) ? helpers_1.Long.fromValue(object.updateGlobalPriceInterval) : helpers_1.Long.UZERO,
            updatePriceDisallowedDays: (0, helpers_1.isSet)(object.updatePriceDisallowedDays) ? Number(object.updatePriceDisallowedDays) : 0
        };
    },
    toJSON(message) {
        const obj = {};
        message.depositDenom !== undefined && (obj.depositDenom = message.depositDenom);
        message.minDeposit !== undefined && (obj.minDeposit = message.minDeposit);
        message.secondarySpStorePriceRatio !== undefined && (obj.secondarySpStorePriceRatio = message.secondarySpStorePriceRatio);
        message.numOfHistoricalBlocksForMaintenanceRecords !== undefined && (obj.numOfHistoricalBlocksForMaintenanceRecords = (message.numOfHistoricalBlocksForMaintenanceRecords || helpers_1.Long.ZERO).toString());
        message.maintenanceDurationQuota !== undefined && (obj.maintenanceDurationQuota = (message.maintenanceDurationQuota || helpers_1.Long.ZERO).toString());
        message.numOfLockupBlocksForMaintenance !== undefined && (obj.numOfLockupBlocksForMaintenance = (message.numOfLockupBlocksForMaintenance || helpers_1.Long.ZERO).toString());
        message.updateGlobalPriceInterval !== undefined && (obj.updateGlobalPriceInterval = (message.updateGlobalPriceInterval || helpers_1.Long.UZERO).toString());
        message.updatePriceDisallowedDays !== undefined && (obj.updatePriceDisallowedDays = Math.round(message.updatePriceDisallowedDays));
        return obj;
    },
    fromPartial(object) {
        const message = createBaseParams();
        message.depositDenom = object.depositDenom ?? "";
        message.minDeposit = object.minDeposit ?? "";
        message.secondarySpStorePriceRatio = object.secondarySpStorePriceRatio ?? "";
        message.numOfHistoricalBlocksForMaintenanceRecords = object.numOfHistoricalBlocksForMaintenanceRecords !== undefined && object.numOfHistoricalBlocksForMaintenanceRecords !== null ? helpers_1.Long.fromValue(object.numOfHistoricalBlocksForMaintenanceRecords) : helpers_1.Long.ZERO;
        message.maintenanceDurationQuota = object.maintenanceDurationQuota !== undefined && object.maintenanceDurationQuota !== null ? helpers_1.Long.fromValue(object.maintenanceDurationQuota) : helpers_1.Long.ZERO;
        message.numOfLockupBlocksForMaintenance = object.numOfLockupBlocksForMaintenance !== undefined && object.numOfLockupBlocksForMaintenance !== null ? helpers_1.Long.fromValue(object.numOfLockupBlocksForMaintenance) : helpers_1.Long.ZERO;
        message.updateGlobalPriceInterval = object.updateGlobalPriceInterval !== undefined && object.updateGlobalPriceInterval !== null ? helpers_1.Long.fromValue(object.updateGlobalPriceInterval) : helpers_1.Long.UZERO;
        message.updatePriceDisallowedDays = object.updatePriceDisallowedDays ?? 0;
        return message;
    },
    fromSDK(object) {
        return {
            depositDenom: object?.deposit_denom,
            minDeposit: object?.min_deposit,
            secondarySpStorePriceRatio: object?.secondary_sp_store_price_ratio,
            numOfHistoricalBlocksForMaintenanceRecords: object?.num_of_historical_blocks_for_maintenance_records,
            maintenanceDurationQuota: object?.maintenance_duration_quota,
            numOfLockupBlocksForMaintenance: object?.num_of_lockup_blocks_for_maintenance,
            updateGlobalPriceInterval: object?.update_global_price_interval,
            updatePriceDisallowedDays: object?.update_price_disallowed_days
        };
    },
    toSDK(message) {
        const obj = {};
        obj.deposit_denom = message.depositDenom;
        obj.min_deposit = message.minDeposit;
        obj.secondary_sp_store_price_ratio = message.secondarySpStorePriceRatio;
        obj.num_of_historical_blocks_for_maintenance_records = message.numOfHistoricalBlocksForMaintenanceRecords;
        obj.maintenance_duration_quota = message.maintenanceDurationQuota;
        obj.num_of_lockup_blocks_for_maintenance = message.numOfLockupBlocksForMaintenance;
        obj.update_global_price_interval = message.updateGlobalPriceInterval;
        obj.update_price_disallowed_days = message.updatePriceDisallowedDays;
        return obj;
    }
};
