// @bnb-chain/greenfield-cosmos-types@0.4.0-alpha.23/greenfield/storage/params.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VersionedParams = exports.Params = exports.protobufPackage = void 0;
//@ts-nocheck
/* eslint-disable */
const helpers_1 = require("../../helpers");
const _m0 = __importStar(require("protobufjs/minimal"));
exports.protobufPackage = "greenfield.storage";
function createBaseParams() {
    return {
        versionedParams: exports.VersionedParams.fromPartial({}),
        maxPayloadSize: helpers_1.Long.UZERO,
        bscMirrorBucketRelayerFee: "",
        bscMirrorBucketAckRelayerFee: "",
        bscMirrorObjectRelayerFee: "",
        bscMirrorObjectAckRelayerFee: "",
        bscMirrorGroupRelayerFee: "",
        bscMirrorGroupAckRelayerFee: "",
        maxBucketsPerAccount: 0,
        discontinueCountingWindow: helpers_1.Long.UZERO,
        discontinueObjectMax: helpers_1.Long.UZERO,
        discontinueBucketMax: helpers_1.Long.UZERO,
        discontinueConfirmPeriod: helpers_1.Long.ZERO,
        discontinueDeletionMax: helpers_1.Long.UZERO,
        stalePolicyCleanupMax: helpers_1.Long.UZERO,
        minQuotaUpdateInterval: helpers_1.Long.UZERO,
        maxLocalVirtualGroupNumPerBucket: 0,
        opMirrorBucketRelayerFee: "",
        opMirrorBucketAckRelayerFee: "",
        opMirrorObjectRelayerFee: "",
        opMirrorObjectAckRelayerFee: "",
        opMirrorGroupRelayerFee: "",
        opMirrorGroupAckRelayerFee: ""
    };
}
exports.Params = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.versionedParams !== undefined) {
            exports.VersionedParams.encode(message.versionedParams, writer.uint32(10).fork()).ldelim();
        }
        if (!message.maxPayloadSize.isZero()) {
            writer.uint32(16).uint64(message.maxPayloadSize);
        }
        if (message.bscMirrorBucketRelayerFee !== "") {
            writer.uint32(26).string(message.bscMirrorBucketRelayerFee);
        }
        if (message.bscMirrorBucketAckRelayerFee !== "") {
            writer.uint32(34).string(message.bscMirrorBucketAckRelayerFee);
        }
        if (message.bscMirrorObjectRelayerFee !== "") {
            writer.uint32(42).string(message.bscMirrorObjectRelayerFee);
        }
        if (message.bscMirrorObjectAckRelayerFee !== "") {
            writer.uint32(50).string(message.bscMirrorObjectAckRelayerFee);
        }
        if (message.bscMirrorGroupRelayerFee !== "") {
            writer.uint32(58).string(message.bscMirrorGroupRelayerFee);
        }
        if (message.bscMirrorGroupAckRelayerFee !== "") {
            writer.uint32(66).string(message.bscMirrorGroupAckRelayerFee);
        }
        if (message.maxBucketsPerAccount !== 0) {
            writer.uint32(72).uint32(message.maxBucketsPerAccount);
        }
        if (!message.discontinueCountingWindow.isZero()) {
            writer.uint32(80).uint64(message.discontinueCountingWindow);
        }
        if (!message.discontinueObjectMax.isZero()) {
            writer.uint32(88).uint64(message.discontinueObjectMax);
        }
        if (!message.discontinueBucketMax.isZero()) {
            writer.uint32(96).uint64(message.discontinueBucketMax);
        }
        if (!message.discontinueConfirmPeriod.isZero()) {
            writer.uint32(104).int64(message.discontinueConfirmPeriod);
        }
        if (!message.discontinueDeletionMax.isZero()) {
            writer.uint32(112).uint64(message.discontinueDeletionMax);
        }
        if (!message.stalePolicyCleanupMax.isZero()) {
            writer.uint32(120).uint64(message.stalePolicyCleanupMax);
        }
        if (!message.minQuotaUpdateInterval.isZero()) {
            writer.uint32(128).uint64(message.minQuotaUpdateInterval);
        }
        if (message.maxLocalVirtualGroupNumPerBucket !== 0) {
            writer.uint32(136).uint32(message.maxLocalVirtualGroupNumPerBucket);
        }
        if (message.opMirrorBucketRelayerFee !== "") {
            writer.uint32(146).string(message.opMirrorBucketRelayerFee);
        }
        if (message.opMirrorBucketAckRelayerFee !== "") {
            writer.uint32(154).string(message.opMirrorBucketAckRelayerFee);
        }
        if (message.opMirrorObjectRelayerFee !== "") {
            writer.uint32(162).string(message.opMirrorObjectRelayerFee);
        }
        if (message.opMirrorObjectAckRelayerFee !== "") {
            writer.uint32(170).string(message.opMirrorObjectAckRelayerFee);
        }
        if (message.opMirrorGroupRelayerFee !== "") {
            writer.uint32(178).string(message.opMirrorGroupRelayerFee);
        }
        if (message.opMirrorGroupAckRelayerFee !== "") {
            writer.uint32(186).string(message.opMirrorGroupAckRelayerFee);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.versionedParams = exports.VersionedParams.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.maxPayloadSize = reader.uint64();
                    break;
                case 3:
                    message.bscMirrorBucketRelayerFee = reader.string();
                    break;
                case 4:
                    message.bscMirrorBucketAckRelayerFee = reader.string();
                    break;
                case 5:
                    message.bscMirrorObjectRelayerFee = reader.string();
                    break;
                case 6:
                    message.bscMirrorObjectAckRelayerFee = reader.string();
                    break;
                case 7:
                    message.bscMirrorGroupRelayerFee = reader.string();
                    break;
                case 8:
                    message.bscMirrorGroupAckRelayerFee = reader.string();
                    break;
                case 9:
                    message.maxBucketsPerAccount = reader.uint32();
                    break;
                case 10:
                    message.discontinueCountingWindow = reader.uint64();
                    break;
                case 11:
                    message.discontinueObjectMax = reader.uint64();
                    break;
                case 12:
                    message.discontinueBucketMax = reader.uint64();
                    break;
                case 13:
                    message.discontinueConfirmPeriod = reader.int64();
                    break;
                case 14:
                    message.discontinueDeletionMax = reader.uint64();
                    break;
                case 15:
                    message.stalePolicyCleanupMax = reader.uint64();
                    break;
                case 16:
                    message.minQuotaUpdateInterval = reader.uint64();
                    break;
                case 17:
                    message.maxLocalVirtualGroupNumPerBucket = reader.uint32();
                    break;
                case 18:
                    message.opMirrorBucketRelayerFee = reader.string();
                    break;
                case 19:
                    message.opMirrorBucketAckRelayerFee = reader.string();
                    break;
                case 20:
                    message.opMirrorObjectRelayerFee = reader.string();
                    break;
                case 21:
                    message.opMirrorObjectAckRelayerFee = reader.string();
                    break;
                case 22:
                    message.opMirrorGroupRelayerFee = reader.string();
                    break;
                case 23:
                    message.opMirrorGroupAckRelayerFee = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            versionedParams: (0, helpers_1.isSet)(object.versionedParams) ? exports.VersionedParams.fromJSON(object.versionedParams) : undefined,
            maxPayloadSize: (0, helpers_1.isSet)(object.maxPayloadSize) ? helpers_1.Long.fromValue(object.maxPayloadSize) : helpers_1.Long.UZERO,
            bscMirrorBucketRelayerFee: (0, helpers_1.isSet)(object.bscMirrorBucketRelayerFee) ? String(object.bscMirrorBucketRelayerFee) : "",
            bscMirrorBucketAckRelayerFee: (0, helpers_1.isSet)(object.bscMirrorBucketAckRelayerFee) ? String(object.bscMirrorBucketAckRelayerFee) : "",
            bscMirrorObjectRelayerFee: (0, helpers_1.isSet)(object.bscMirrorObjectRelayerFee) ? String(object.bscMirrorObjectRelayerFee) : "",
            bscMirrorObjectAckRelayerFee: (0, helpers_1.isSet)(object.bscMirrorObjectAckRelayerFee) ? String(object.bscMirrorObjectAckRelayerFee) : "",
            bscMirrorGroupRelayerFee: (0, helpers_1.isSet)(object.bscMirrorGroupRelayerFee) ? String(object.bscMirrorGroupRelayerFee) : "",
            bscMirrorGroupAckRelayerFee: (0, helpers_1.isSet)(object.bscMirrorGroupAckRelayerFee) ? String(object.bscMirrorGroupAckRelayerFee) : "",
            maxBucketsPerAccount: (0, helpers_1.isSet)(object.maxBucketsPerAccount) ? Number(object.maxBucketsPerAccount) : 0,
            discontinueCountingWindow: (0, helpers_1.isSet)(object.discontinueCountingWindow) ? helpers_1.Long.fromValue(object.discontinueCountingWindow) : helpers_1.Long.UZERO,
            discontinueObjectMax: (0, helpers_1.isSet)(object.discontinueObjectMax) ? helpers_1.Long.fromValue(object.discontinueObjectMax) : helpers_1.Long.UZERO,
            discontinueBucketMax: (0, helpers_1.isSet)(object.discontinueBucketMax) ? helpers_1.Long.fromValue(object.discontinueBucketMax) : helpers_1.Long.UZERO,
            discontinueConfirmPeriod: (0, helpers_1.isSet)(object.discontinueConfirmPeriod) ? helpers_1.Long.fromValue(object.discontinueConfirmPeriod) : helpers_1.Long.ZERO,
            discontinueDeletionMax: (0, helpers_1.isSet)(object.discontinueDeletionMax) ? helpers_1.Long.fromValue(object.discontinueDeletionMax) : helpers_1.Long.UZERO,
            stalePolicyCleanupMax: (0, helpers_1.isSet)(object.stalePolicyCleanupMax) ? helpers_1.Long.fromValue(object.stalePolicyCleanupMax) : helpers_1.Long.UZERO,
            minQuotaUpdateInterval: (0, helpers_1.isSet)(object.minQuotaUpdateInterval) ? helpers_1.Long.fromValue(object.minQuotaUpdateInterval) : helpers_1.Long.UZERO,
            maxLocalVirtualGroupNumPerBucket: (0, helpers_1.isSet)(object.maxLocalVirtualGroupNumPerBucket) ? Number(object.maxLocalVirtualGroupNumPerBucket) : 0,
            opMirrorBucketRelayerFee: (0, helpers_1.isSet)(object.opMirrorBucketRelayerFee) ? String(object.opMirrorBucketRelayerFee) : "",
            opMirrorBucketAckRelayerFee: (0, helpers_1.isSet)(object.opMirrorBucketAckRelayerFee) ? String(object.opMirrorBucketAckRelayerFee) : "",
            opMirrorObjectRelayerFee: (0, helpers_1.isSet)(object.opMirrorObjectRelayerFee) ? String(object.opMirrorObjectRelayerFee) : "",
            opMirrorObjectAckRelayerFee: (0, helpers_1.isSet)(object.opMirrorObjectAckRelayerFee) ? String(object.opMirrorObjectAckRelayerFee) : "",
            opMirrorGroupRelayerFee: (0, helpers_1.isSet)(object.opMirrorGroupRelayerFee) ? String(object.opMirrorGroupRelayerFee) : "",
            opMirrorGroupAckRelayerFee: (0, helpers_1.isSet)(object.opMirrorGroupAckRelayerFee) ? String(object.opMirrorGroupAckRelayerFee) : ""
        };
    },
    toJSON(message) {
        const obj = {};
        message.versionedParams !== undefined && (obj.versionedParams = message.versionedParams ? exports.VersionedParams.toJSON(message.versionedParams) : undefined);
        message.maxPayloadSize !== undefined && (obj.maxPayloadSize = (message.maxPayloadSize || helpers_1.Long.UZERO).toString());
        message.bscMirrorBucketRelayerFee !== undefined && (obj.bscMirrorBucketRelayerFee = message.bscMirrorBucketRelayerFee);
        message.bscMirrorBucketAckRelayerFee !== undefined && (obj.bscMirrorBucketAckRelayerFee = message.bscMirrorBucketAckRelayerFee);
        message.bscMirrorObjectRelayerFee !== undefined && (obj.bscMirrorObjectRelayerFee = message.bscMirrorObjectRelayerFee);
        message.bscMirrorObjectAckRelayerFee !== undefined && (obj.bscMirrorObjectAckRelayerFee = message.bscMirrorObjectAckRelayerFee);
        message.bscMirrorGroupRelayerFee !== undefined && (obj.bscMirrorGroupRelayerFee = message.bscMirrorGroupRelayerFee);
        message.bscMirrorGroupAckRelayerFee !== undefined && (obj.bscMirrorGroupAckRelayerFee = message.bscMirrorGroupAckRelayerFee);
        message.maxBucketsPerAccount !== undefined && (obj.maxBucketsPerAccount = Math.round(message.maxBucketsPerAccount));
        message.discontinueCountingWindow !== undefined && (obj.discontinueCountingWindow = (message.discontinueCountingWindow || helpers_1.Long.UZERO).toString());
        message.discontinueObjectMax !== undefined && (obj.discontinueObjectMax = (message.discontinueObjectMax || helpers_1.Long.UZERO).toString());
        message.discontinueBucketMax !== undefined && (obj.discontinueBucketMax = (message.discontinueBucketMax || helpers_1.Long.UZERO).toString());
        message.discontinueConfirmPeriod !== undefined && (obj.discontinueConfirmPeriod = (message.discontinueConfirmPeriod || helpers_1.Long.ZERO).toString());
        message.discontinueDeletionMax !== undefined && (obj.discontinueDeletionMax = (message.discontinueDeletionMax || helpers_1.Long.UZERO).toString());
        message.stalePolicyCleanupMax !== undefined && (obj.stalePolicyCleanupMax = (message.stalePolicyCleanupMax || helpers_1.Long.UZERO).toString());
        message.minQuotaUpdateInterval !== undefined && (obj.minQuotaUpdateInterval = (message.minQuotaUpdateInterval || helpers_1.Long.UZERO).toString());
        message.maxLocalVirtualGroupNumPerBucket !== undefined && (obj.maxLocalVirtualGroupNumPerBucket = Math.round(message.maxLocalVirtualGroupNumPerBucket));
        message.opMirrorBucketRelayerFee !== undefined && (obj.opMirrorBucketRelayerFee = message.opMirrorBucketRelayerFee);
        message.opMirrorBucketAckRelayerFee !== undefined && (obj.opMirrorBucketAckRelayerFee = message.opMirrorBucketAckRelayerFee);
        message.opMirrorObjectRelayerFee !== undefined && (obj.opMirrorObjectRelayerFee = message.opMirrorObjectRelayerFee);
        message.opMirrorObjectAckRelayerFee !== undefined && (obj.opMirrorObjectAckRelayerFee = message.opMirrorObjectAckRelayerFee);
        message.opMirrorGroupRelayerFee !== undefined && (obj.opMirrorGroupRelayerFee = message.opMirrorGroupRelayerFee);
        message.opMirrorGroupAckRelayerFee !== undefined && (obj.opMirrorGroupAckRelayerFee = message.opMirrorGroupAckRelayerFee);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseParams();
        message.versionedParams = object.versionedParams !== undefined && object.versionedParams !== null ? exports.VersionedParams.fromPartial(object.versionedParams) : undefined;
        message.maxPayloadSize = object.maxPayloadSize !== undefined && object.maxPayloadSize !== null ? helpers_1.Long.fromValue(object.maxPayloadSize) : helpers_1.Long.UZERO;
        message.bscMirrorBucketRelayerFee = object.bscMirrorBucketRelayerFee ?? "";
        message.bscMirrorBucketAckRelayerFee = object.bscMirrorBucketAckRelayerFee ?? "";
        message.bscMirrorObjectRelayerFee = object.bscMirrorObjectRelayerFee ?? "";
        message.bscMirrorObjectAckRelayerFee = object.bscMirrorObjectAckRelayerFee ?? "";
        message.bscMirrorGroupRelayerFee = object.bscMirrorGroupRelayerFee ?? "";
        message.bscMirrorGroupAckRelayerFee = object.bscMirrorGroupAckRelayerFee ?? "";
        message.maxBucketsPerAccount = object.maxBucketsPerAccount ?? 0;
        message.discontinueCountingWindow = object.discontinueCountingWindow !== undefined && object.discontinueCountingWindow !== null ? helpers_1.Long.fromValue(object.discontinueCountingWindow) : helpers_1.Long.UZERO;
        message.discontinueObjectMax = object.discontinueObjectMax !== undefined && object.discontinueObjectMax !== null ? helpers_1.Long.fromValue(object.discontinueObjectMax) : helpers_1.Long.UZERO;
        message.discontinueBucketMax = object.discontinueBucketMax !== undefined && object.discontinueBucketMax !== null ? helpers_1.Long.fromValue(object.discontinueBucketMax) : helpers_1.Long.UZERO;
        message.discontinueConfirmPeriod = object.discontinueConfirmPeriod !== undefined && object.discontinueConfirmPeriod !== null ? helpers_1.Long.fromValue(object.discontinueConfirmPeriod) : helpers_1.Long.ZERO;
        message.discontinueDeletionMax = object.discontinueDeletionMax !== undefined && object.discontinueDeletionMax !== null ? helpers_1.Long.fromValue(object.discontinueDeletionMax) : helpers_1.Long.UZERO;
        message.stalePolicyCleanupMax = object.stalePolicyCleanupMax !== undefined && object.stalePolicyCleanupMax !== null ? helpers_1.Long.fromValue(object.stalePolicyCleanupMax) : helpers_1.Long.UZERO;
        message.minQuotaUpdateInterval = object.minQuotaUpdateInterval !== undefined && object.minQuotaUpdateInterval !== null ? helpers_1.Long.fromValue(object.minQuotaUpdateInterval) : helpers_1.Long.UZERO;
        message.maxLocalVirtualGroupNumPerBucket = object.maxLocalVirtualGroupNumPerBucket ?? 0;
        message.opMirrorBucketRelayerFee = object.opMirrorBucketRelayerFee ?? "";
        message.opMirrorBucketAckRelayerFee = object.opMirrorBucketAckRelayerFee ?? "";
        message.opMirrorObjectRelayerFee = object.opMirrorObjectRelayerFee ?? "";
        message.opMirrorObjectAckRelayerFee = object.opMirrorObjectAckRelayerFee ?? "";
        message.opMirrorGroupRelayerFee = object.opMirrorGroupRelayerFee ?? "";
        message.opMirrorGroupAckRelayerFee = object.opMirrorGroupAckRelayerFee ?? "";
        return message;
    },
    fromSDK(object) {
        return {
            versionedParams: object.versioned_params ? exports.VersionedParams.fromSDK(object.versioned_params) : undefined,
            maxPayloadSize: object?.max_payload_size,
            bscMirrorBucketRelayerFee: object?.bsc_mirror_bucket_relayer_fee,
            bscMirrorBucketAckRelayerFee: object?.bsc_mirror_bucket_ack_relayer_fee,
            bscMirrorObjectRelayerFee: object?.bsc_mirror_object_relayer_fee,
            bscMirrorObjectAckRelayerFee: object?.bsc_mirror_object_ack_relayer_fee,
            bscMirrorGroupRelayerFee: object?.bsc_mirror_group_relayer_fee,
            bscMirrorGroupAckRelayerFee: object?.bsc_mirror_group_ack_relayer_fee,
            maxBucketsPerAccount: object?.max_buckets_per_account,
            discontinueCountingWindow: object?.discontinue_counting_window,
            discontinueObjectMax: object?.discontinue_object_max,
            discontinueBucketMax: object?.discontinue_bucket_max,
            discontinueConfirmPeriod: object?.discontinue_confirm_period,
            discontinueDeletionMax: object?.discontinue_deletion_max,
            stalePolicyCleanupMax: object?.stale_policy_cleanup_max,
            minQuotaUpdateInterval: object?.min_quota_update_interval,
            maxLocalVirtualGroupNumPerBucket: object?.max_local_virtual_group_num_per_bucket,
            opMirrorBucketRelayerFee: object?.op_mirror_bucket_relayer_fee,
            opMirrorBucketAckRelayerFee: object?.op_mirror_bucket_ack_relayer_fee,
            opMirrorObjectRelayerFee: object?.op_mirror_object_relayer_fee,
            opMirrorObjectAckRelayerFee: object?.op_mirror_object_ack_relayer_fee,
            opMirrorGroupRelayerFee: object?.op_mirror_group_relayer_fee,
            opMirrorGroupAckRelayerFee: object?.op_mirror_group_ack_relayer_fee
        };
    },
    toSDK(message) {
        const obj = {};
        message.versionedParams !== undefined && (obj.versioned_params = message.versionedParams ? exports.VersionedParams.toSDK(message.versionedParams) : undefined);
        obj.max_payload_size = message.maxPayloadSize;
        obj.bsc_mirror_bucket_relayer_fee = message.bscMirrorBucketRelayerFee;
        obj.bsc_mirror_bucket_ack_relayer_fee = message.bscMirrorBucketAckRelayerFee;
        obj.bsc_mirror_object_relayer_fee = message.bscMirrorObjectRelayerFee;
        obj.bsc_mirror_object_ack_relayer_fee = message.bscMirrorObjectAckRelayerFee;
        obj.bsc_mirror_group_relayer_fee = message.bscMirrorGroupRelayerFee;
        obj.bsc_mirror_group_ack_relayer_fee = message.bscMirrorGroupAckRelayerFee;
        obj.max_buckets_per_account = message.maxBucketsPerAccount;
        obj.discontinue_counting_window = message.discontinueCountingWindow;
        obj.discontinue_object_max = message.discontinueObjectMax;
        obj.discontinue_bucket_max = message.discontinueBucketMax;
        obj.discontinue_confirm_period = message.discontinueConfirmPeriod;
        obj.discontinue_deletion_max = message.discontinueDeletionMax;
        obj.stale_policy_cleanup_max = message.stalePolicyCleanupMax;
        obj.min_quota_update_interval = message.minQuotaUpdateInterval;
        obj.max_local_virtual_group_num_per_bucket = message.maxLocalVirtualGroupNumPerBucket;
        obj.op_mirror_bucket_relayer_fee = message.opMirrorBucketRelayerFee;
        obj.op_mirror_bucket_ack_relayer_fee = message.opMirrorBucketAckRelayerFee;
        obj.op_mirror_object_relayer_fee = message.opMirrorObjectRelayerFee;
        obj.op_mirror_object_ack_relayer_fee = message.opMirrorObjectAckRelayerFee;
        obj.op_mirror_group_relayer_fee = message.opMirrorGroupRelayerFee;
        obj.op_mirror_group_ack_relayer_fee = message.opMirrorGroupAckRelayerFee;
        return obj;
    }
};
function createBaseVersionedParams() {
    return {
        maxSegmentSize: helpers_1.Long.UZERO,
        redundantDataChunkNum: 0,
        redundantParityChunkNum: 0,
        minChargeSize: helpers_1.Long.UZERO
    };
}
exports.VersionedParams = {
    encode(message, writer = _m0.Writer.create()) {
        if (!message.maxSegmentSize.isZero()) {
            writer.uint32(8).uint64(message.maxSegmentSize);
        }
        if (message.redundantDataChunkNum !== 0) {
            writer.uint32(16).uint32(message.redundantDataChunkNum);
        }
        if (message.redundantParityChunkNum !== 0) {
            writer.uint32(24).uint32(message.redundantParityChunkNum);
        }
        if (!message.minChargeSize.isZero()) {
            writer.uint32(32).uint64(message.minChargeSize);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseVersionedParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.maxSegmentSize = reader.uint64();
                    break;
                case 2:
                    message.redundantDataChunkNum = reader.uint32();
                    break;
                case 3:
                    message.redundantParityChunkNum = reader.uint32();
                    break;
                case 4:
                    message.minChargeSize = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            maxSegmentSize: (0, helpers_1.isSet)(object.maxSegmentSize) ? helpers_1.Long.fromValue(object.maxSegmentSize) : helpers_1.Long.UZERO,
            redundantDataChunkNum: (0, helpers_1.isSet)(object.redundantDataChunkNum) ? Number(object.redundantDataChunkNum) : 0,
            redundantParityChunkNum: (0, helpers_1.isSet)(object.redundantParityChunkNum) ? Number(object.redundantParityChunkNum) : 0,
            minChargeSize: (0, helpers_1.isSet)(object.minChargeSize) ? helpers_1.Long.fromValue(object.minChargeSize) : helpers_1.Long.UZERO
        };
    },
    toJSON(message) {
        const obj = {};
        message.maxSegmentSize !== undefined && (obj.maxSegmentSize = (message.maxSegmentSize || helpers_1.Long.UZERO).toString());
        message.redundantDataChunkNum !== undefined && (obj.redundantDataChunkNum = Math.round(message.redundantDataChunkNum));
        message.redundantParityChunkNum !== undefined && (obj.redundantParityChunkNum = Math.round(message.redundantParityChunkNum));
        message.minChargeSize !== undefined && (obj.minChargeSize = (message.minChargeSize || helpers_1.Long.UZERO).toString());
        return obj;
    },
    fromPartial(object) {
        const message = createBaseVersionedParams();
        message.maxSegmentSize = object.maxSegmentSize !== undefined && object.maxSegmentSize !== null ? helpers_1.Long.fromValue(object.maxSegmentSize) : helpers_1.Long.UZERO;
        message.redundantDataChunkNum = object.redundantDataChunkNum ?? 0;
        message.redundantParityChunkNum = object.redundantParityChunkNum ?? 0;
        message.minChargeSize = object.minChargeSize !== undefined && object.minChargeSize !== null ? helpers_1.Long.fromValue(object.minChargeSize) : helpers_1.Long.UZERO;
        return message;
    },
    fromSDK(object) {
        return {
            maxSegmentSize: object?.max_segment_size,
            redundantDataChunkNum: object?.redundant_data_chunk_num,
            redundantParityChunkNum: object?.redundant_parity_chunk_num,
            minChargeSize: object?.min_charge_size
        };
    },
    toSDK(message) {
        const obj = {};
        obj.max_segment_size = message.maxSegmentSize;
        obj.redundant_data_chunk_num = message.redundantDataChunkNum;
        obj.redundant_parity_chunk_num = message.redundantParityChunkNum;
        obj.min_charge_size = message.minChargeSize;
        return obj;
    }
};
